### ISSUE TITLE
Token Substitution Support for Check Source

### ISSUE THREAD

**Issue by Evesy** [sensu/sensu#1634]
It would be great if token substitution could be extended to the source attribute in checks, allowing for more dynamic proxy clients, based on certain client attributes.
Token substitution is incredibly powerful and useful, so it'd be nice to extend the functionality to wherever it has a use case.

## Expected Behavior
Using a check token in the 'source' field of a check should function just as it does in others, such as the check attribute. The token should be evaluated against the respective client attribute and a proxy client created accordingly.

## Current Behavior
The source field is currently validated against strict regex that does not allow the use of colons. However as per https://github.com/sensu/sensu/issues/1281#issuecomment-222533017 it appears it is not as simple as just adjusting that validation.

## Possible Solution
To my eyes it appears relaxing the validation would allow token substitution to take place and function as expected.
Unmatched tokens could be avoided perhaps by forcing the use of a default value, or rather having Sensu fall back to a hardcoded default if substitution for the source field fails, however I'm sure there's more to it than that.

## Context
We would like to be able to run a check across several different Elasticsearch clusters (That make use of a proxy client to give a cluster overview (Status, allocation etc.)) and have it all defined via just one check definition, that can make use of client attributes to generate the proxy client name. Example below:

```
{
  "checks": {
    "elasticsearch_cluster_status": {
      "command": "/opt/sensu/embedded/bin/check-es-cluster-status.rb -h localhost -p 9200",
      "subscribers": [
        "roundrobin:elasticsearch",
        "roundrobin:elasticsearch-monitoring"
      ],
      "interval": 60,
      "ttl": 90,
      "occurrences": 5,
      "refresh": 3600,
      "source": "elastic-:::cluster_name:::"
    }
  }
}
```

**Comment by portertech** [comment]
My primary concern is that this functionality can introduce special characters into one or more proxy client names in the registry, which will cause problems when querying the API etc.

**Comment by portertech** [comment]
We should revisit this for 1.1

**Comment by mcbsdEU** [comment]
Don't proxy requests aim for this use-case? You can do tokens in command and source will be automatically filled by proxy client name....or?

**Comment by Evesy** [comment]
> Don't proxy requests aim for this use-case? You can do tokens in command and source will be automatically filled by proxy client name....or?

Proxy requests would certainly work for a number of the checks we have, namely those that can be ran remotely against a host/API endpoint, which does cover a lot of them. However we still have a number of checks that need to/benefit from running the check directly on one of the specific nodes, which proxy requests do not allow (As I'd need the check to be ran on different subscribers depending on the proxy client it's running for)

Even without the above consideration there's still the need to have a means to register the proxy client in the first instance, although this could be achieved quite simply within our chef cookbook, so not a massive deal. 
Doing that however takes away the idea of being able to dynamically add new clusters to sensu -- Rather than it being as simple as setting up the relevant client config, and then adding an additional subscriber to the relevant checks (Which is very straightforward, even for someone not too familiar with Sensu), there'd be the need for an additional step to also create a client in the Sensu registry.

It's possible to achieve what we're after right now, by simply creating essentially duplicate checks, however I don't believe it's the best approach to be taking.

**Comment by Evesy** [comment]
Might be out of scope but I'd also love if this was extended to also support token substitution for named aggregates.

**Context**
Multiple Kafka clusters in a single Sensu datacenter.

I want to run aggregates for Kafka checks that target these environments. In order to do this currently I need to create two separate check definitions, one that targets one set of kafka nodes, and one that targets the other (using two different subscriptions), and then set different aggregate names for those checks.

I'd like to be able to target these checks against a generic subscription (i.e. kafka), and then aggregate based on attribute, i.e. `"aggregate": "kafka_node_health_:::environment:::"`

**Comment by portertech** [comment]
@Evesy it seems validation already allows `:` and `|`, is it not already possible? https://github.com/sensu/sensu-settings/blob/a43e626184ff94adebe2c6c6df710ef33043cbee/lib/sensu/settings/validators/check.rb#L149-L150

**Comment by portertech** [comment]
@Evesy I just tested `"aggregates": ["foo", ":::name|bar:::"]` and had success :)

**Comment by Evesy** [comment]
@portertech Oh, awesome! I'll admit I never actually tried, I'd just assumed since source wasn't supported, aggregate wouldn't be either.

I'll give myself a slap on the wrist....

**Comment by portertech** [comment]
Closed by #1791
